config/internal/configsource/manager.go

```diff
@@ -457,21 +491,54 @@ func (m *Manager) expandString(ctx context.Context, s string) (interface{}, erro
 	return string(buf) + s[i:], nil
 }
 
-func (m *Manager) retrieveConfigSourceData(ctx context.Context, name, cfgSrcInvoke string) (interface{}, error) {
-	cfgSrc, ok := m.configSources[name]
+// retrieveConfigSourceData retrieves data from the specified config source and injects them into
+// the configuration. The Manager tracks sessions and watcher objects as needed.
+func (m *Manager) retrieveConfigSourceData(ctx context.Context, cfgSrcName, cfgSrcInvocation string) (interface{}, error) {
+	cfgSrc, ok := m.configSources[cfgSrcName]
 	if !ok {
-		return nil, newErrUnknownConfigSource(name)
+		return nil, newErrUnknownConfigSource(cfgSrcName)
 	}
 
-	// Expand any env vars on the selector and parameters. Nested config source usage
-	// is not supported.
-	cfgSrcInvoke = expandEnvVars(cfgSrcInvoke)
-	retrieved, err := m.expandConfigSource(ctx, cfgSrc, cfgSrcInvoke)
+	cfgSrcName, selector, paramsParser, err := parseCfgSrcInvocation(cfgSrcInvocation)
 	if err != nil {
 		return nil, err
 	}
 
-	return retrieved, nil
+	session, ok := m.sessions[cfgSrcName]
+	if !ok {
+		session, err = cfgSrc.NewSession(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create session for config source %q: %w", cfgSrcName, err)
+		}
+		m.sessions[cfgSrcName] = session
+	}
+
+	// Recursively expand the selector.
+	var expandedSelector interface{}
+	expandedSelector, err = m.parseStringValue(ctx, selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to process selector for config source %q selector %q: %w", cfgSrcName, selector, err)
+	}
+	if selector, ok = expandedSelector.(string); !ok {
+		return nil, fmt.Errorf("processed selector must be a string instead got a %T %v", expandedSelector, expandedSelector)
+	}
+
+	// Recursively resolve/parse any config source on the parameters.
+	if paramsParser != nil {
+		paramsParser, err = m.Resolve(ctx, paramsParser)
+		if err != nil {
+			return nil, fmt.Errorf("failed to process parameters for config source %q invocation %q: %w", cfgSrcName, cfgSrcInvocation, err)
+		}
+	}
+
+	retrieved, err := session.Retrieve(ctx, selector, paramsParser)
+	if err != nil {
+		return nil, fmt.Errorf("config source %q failed to retrieve value: %w", cfgSrcName, err)
+	}
+
+	m.watchers = append(m.watchers, retrieved.WatchForUpdate)
+
+	return retrieved.Value(), nil
```

Comment:
Shouldn't retrieved value be recursively expanded again before we return it to the caller? Is it not what is necessary for nested config sources to work? (which I am guessing is the goal of this PR)

Reply:
The recursion is only at the parameters. i.e. a config source `Retrieve` operation can have other config source used to obtain the actual retrieve arguments. It is not necessary, at this point, to make it recursive for the data returned.